File: app/src/main/java/maderski/bluetoothautoplaymusic/Helpers/LaunchAppHelper.java
package maderski.bluetoothautoplaymusic.Helpers;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.StringDef;
import android.util.Log;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import maderski.bluetoothautoplaymusic.PackageTools;
import maderski.bluetoothautoplaymusic.SharedPrefs.BAPMPreferences;
import maderski.bluetoothautoplaymusic.UI.activities.LaunchBAPMActivity;
import maderski.bluetoothautoplaymusic.UI.activities.MainActivity;

/**
 * Created by Jason on 12/8/15.
 */
public class LaunchAppHelper extends PackageTools {

    private static final String TAG = LaunchAppHelper.class.getName();

    @StringDef({
            DirectionLocations.HOME,
            DirectionLocations.WORK,
            DirectionLocations.CUSTOM
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface DirectionLocations {
        String HOME = "Home";
        String WORK = "Work";
        String CUSTOM = "Custom";
    }

    private String mDirectionLocation;
    private List<String> mCanLaunchThisTimeLocations = new ArrayList<>();

    public LaunchAppHelper(){
        super();
    }

    //Create a delay before the Music App is launched and if enable launchPackage maps
    public void musicPlayerLaunch(final Context context, int seconds){
        final String pkgName = BAPMPreferences.getPkgSelectedMusicPlayer(context);
        seconds = seconds * 1000;
        Handler handler = new Handler();
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                launchPackage(context, pkgName);
            }
        };
        handler.postDelayed(runnable, seconds);
    }

    //Launch Maps or Waze with a delay
    public void launchMaps(final Context context, int seconds){
        final boolean canLaunchDirections = BAPMPreferences.getCanLaunchDirections(context);
        boolean canLaunchMapsNow = canMapsLaunchNow(context);

        if(canLaunchMapsNow) {
            seconds = seconds * 1000;

            Handler handler = new Handler();
            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    String mapAppName = BAPMPreferences.getMapsChoice(context);
                    if(canLaunchDirections){
                        Uri data = getMapsChoiceUri(context);
                        launchPackage(context, mapAppName, data, Intent.ACTION_VIEW);
                    } else {
                        determineIfLaunchWithDrivingMode(context, mapAppName);
                    }
                    Log.d(TAG, "delayLaunchmaps started");
                }
            };
            handler.postDelayed(runnable, seconds);
        }
    }

    // If driving mode is enabled and map choice is set to Google Maps, launch Maps in Driving Mode
    private void determineIfLaunchWithDrivingMode(final Context context, final String mapAppName) {
        boolean canLaunchDrivingMode = BAPMPreferences.getLaunchMapsDrivingMode(context) &&
                mapAppName.equals(PackageName.MAPS);
        if(canLaunchDrivingMode){
            Log.d(TAG, "LAUNCH DRIVING MODE");
            Uri data = Uri.parse("google.navigation:/?free=1&mode=d&entry=fnls");
            launchPackage(context, mapAppName, data, Intent.ACTION_VIEW);
        } else {
            launchPackage(context, mapAppName);
        }
    }

    private Uri getMapsChoiceUri(Context context){
        if(mDirectionLocation != null) {
            mDirectionLocation = mDirectionLocation.equals(DirectionLocations.CUSTOM)
                    ? BAPMPreferences.getCustomLocationName(context) : mDirectionLocation;
        }

        Uri uri;
        if(BAPMPreferences.getMapsChoice(context).equals(PackageName.WAZE)){
            String wazeUri = "waze://?favorite=" + mDirectionLocation + "&navigate=yes";
            uri = Uri.parse(wazeUri);
        } else {
            String mapsUri = "google.navigation:q="+mDirectionLocation;
            uri = Uri.parse(mapsUri);
        }

        Log.d(TAG, "MAPS CHOICE URI:" + uri.toString());
        return uri;
    }

    public void launchBAPMActivity(final Context context){
        Handler handler = new Handler(Looper.getMainLooper());
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                Intent[] intentArray = new Intent[2];
                intentArray[0] = new Intent(context, MainActivity.class);
                intentArray[0].addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                intentArray[1] = new Intent(context, LaunchBAPMActivity.class);
                context.startActivities(intentArray);
            }
        };
        handler.postDelayed(runnable, 750);
    }

    public void sendEverythingToBackground(Context context){
        Intent i = new Intent(Intent.ACTION_MAIN);
        i.addCategory(Intent.CATEGORY_HOME);
        i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(i);
    }

    public boolean canMapsLaunchNow(Context context) {
        final boolean canLaunchDuringThisTime = canLaunchDuringThisTime(context);

        boolean canLaunchMapsNow = false;
        if(canLaunchDuringThisTime) {
            for(String location : mCanLaunchThisTimeLocations) {
                canLaunchMapsNow = canLaunchOnThisDay(context, location);
                if(canLaunchMapsNow) {
                    mDirectionLocation = location;
                    break;
                }
            }
        }
        mCanLaunchThisTimeLocations.clear();

        return canLaunchMapsNow;
    }

    public boolean canLaunchOnThisDay(Context context, @DirectionLocations String directionLocation) {
        Calendar calendar = Calendar.getInstance();
        String today = Integer.toString(calendar.get(Calendar.DAY_OF_WEEK));
        boolean canLaunch = false;

        switch(directionLocation) {
            case DirectionLocations.HOME:
                canLaunch = BAPMPreferences.getHomeDaysToLaunchMaps(context).contains(today);
                break;
            case DirectionLocations.WORK:
                canLaunch = BAPMPreferences.getWorkDaysToLaunchMaps(context).contains(today);
                break;
            case DirectionLocations.CUSTOM:
                canLaunch = BAPMPreferences.getCustomDaysToLaunchMaps(context).contains(today);
                break;
        }

        Log.d(TAG, "Day of the week: " + today);
        Log.d(TAG, "Can Launch: " + canLaunch);
        Log.d(TAG, "Direction Location: " + directionLocation);

        return canLaunch;
    }

    public boolean canLaunchDuringThisTime(Context context){
        boolean isUseLaunchTimeEnabled = BAPMPreferences.getUseTimesToLaunchMaps(context);
        if(isUseLaunchTimeEnabled) {

            int morningStartTime = BAPMPreferences.getMorningStartTime(context);
            int morningEndTime = BAPMPreferences.getMorningEndTime(context);

            int eveningStartTime = BAPMPreferences.getEveningStartTime(context);
            int eveningEndTime = BAPMPreferences.getEveningEndTime(context);

            int customStartTime = BAPMPreferences.getCustomStartTime(context);
            int customEndTime = BAPMPreferences.getCustomEndTime(context);

            int current24hrTime = TimeHelper.getCurrent24hrTime();

            TimeHelper timeHelperMorning = new TimeHelper(morningStartTime, morningEndTime, current24hrTime);
            boolean canLaunchWork = timeHelperMorning.isWithinTimeSpan();
            if(canLaunchWork) {
                mCanLaunchThisTimeLocations.add(DirectionLocations.WORK);
            }

            TimeHelper timeHelperEvening = new TimeHelper(eveningStartTime, eveningEndTime, current24hrTime);
            boolean canLaunchHome = timeHelperEvening.isWithinTimeSpan();
            if(canLaunchHome) {
                mCanLaunchThisTimeLocations.add(DirectionLocations.HOME);
            }

            TimeHelper timeHelperCustom = new TimeHelper(customStartTime, customEndTime, current24hrTime);
            boolean canLaunchCustom = timeHelperCustom.isWithinTimeSpan();
            if(canLaunchCustom) {
                mCanLaunchThisTimeLocations.add(DirectionLocations.CUSTOM);
            }

            return canLaunchWork || canLaunchHome || canLaunchCustom;
        } else {
            return true;
        }
    }

    public void launchWazeDirections(final Context context, final String location){
        Handler handler = new Handler();
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                String uriString = "waze://?favorite=" + location + "&navigate=yes";
                Log.d(TAG, "DIRECTIONS LOCATION: " + uriString);
                Uri uri = Uri.parse(uriString);
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(uri);
                context.sendBroadcast(intent);
            }
        };

        handler.postDelayed(runnable, 4000);
    }

    public void closeWazeOnDisconnect(final Context context){
        Handler handler = new Handler();
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                Intent intent = new Intent();
                intent.setAction("Eliran_Close_Intent");
                context.sendBroadcast(intent);
            }
        };

        handler.postDelayed(runnable, 2000);
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/helpers/LaunchAppHelper.kt
package maderski.bluetoothautoplaymusic.helpers

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import maderski.bluetoothautoplaymusic.helpers.enums.DirectionLocation
import maderski.bluetoothautoplaymusic.helpers.enums.DirectionLocation.*
import maderski.bluetoothautoplaymusic.helpers.enums.MapApps.*
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences
import maderski.bluetoothautoplaymusic.ui.activities.DisconnectActivity
import maderski.bluetoothautoplaymusic.ui.activities.LaunchBAPMActivity
import maderski.bluetoothautoplaymusic.ui.activities.MainActivity
import java.util.*


/**
 * Created by Jason on 12/8/15.
 */
class LaunchAppHelper(
        private val context: Context,
        private val packageHelper: PackageHelper,
        private val preferencesHelper: PreferencesHelper
) {
    private val canLaunchThisTimeLocations = ArrayList<DirectionLocation>()

    private var directionLocation: DirectionLocation = NONE

    fun launchApp(packageName: String) = packageHelper.launchPackage(packageName)

    fun mapsLaunch() {
        val isMapsRunning = packageHelper.isAppRunning(MAPS.packageName)
        val canLaunch = canMapsLaunchNow()
        if (canLaunch && !isMapsRunning) {
            val canLaunchDirections = preferencesHelper.canLaunchDirections
            if (canLaunchDirections) {
                val mapAppName = preferencesHelper.mapAppName
                val data = getMapsChoiceUri()
                packageHelper.launchPackage(mapAppName, data, Intent.ACTION_VIEW)
            } else {
                determineHowToLaunchMaps()
            }
            Log.d(TAG, "Launch maps started")
        }
    }

    // Launch Maps or Waze with a delay
    fun launchMapsDelayed(seconds: Int) {
        val mills = seconds * 1000L

        val handler = Handler(Looper.getMainLooper())
        val runnable = Runnable {
            mapsLaunch()
        }
        handler.postDelayed(runnable, mills)
    }

    private fun determineHowToLaunchMaps() {
        val mapAppName = preferencesHelper.mapAppName
        val canLaunchDirections = preferencesHelper.canLaunchDirections
        if (canLaunchDirections && directionLocation != NONE) {
            val data = getMapsChoiceUri()
            packageHelper.launchPackage(mapAppName, data, Intent.ACTION_VIEW)
        } else {
            determineIfLaunchWithDrivingMode(mapAppName)
        }
    }

    // If driving mode is enabled and map choice is set to Google Maps, launch Maps in Driving Mode
    private fun determineIfLaunchWithDrivingMode(mapAppName: String) {
        val canLaunchDrivingMode = preferencesHelper.canLaunchDrivingMode
        if (canLaunchDrivingMode) {
            Log.d(TAG, "LAUNCH DRIVING MODE")
            val data = Uri.parse("google.navigation:/?free=1&mode=d&entry=fnls")
            packageHelper.launchPackage(mapAppName, data, Intent.ACTION_VIEW)
        } else {
            packageHelper.launchPackage(mapAppName)
        }
    }

    private fun getMapsChoiceUri(): Uri {
        val customLocationName = preferencesHelper.customLocationName
        val location = if (directionLocation == CUSTOM)
            customLocationName
        else
            directionLocation.location
        val mapAppChosen = preferencesHelper.mapAppChosen
        val uri = if (mapAppChosen == WAZE.packageName) {
            val wazeUri = "waze://?favorite=$location&navigate=yes"
            Uri.parse(wazeUri)
        } else {
            val mapsUri = "google.navigation:q=$location"
            Uri.parse(mapsUri)
        }

        Log.d(TAG, "MAPS CHOICE URI:$uri")
        return uri
    }

    fun launchBAPMActivity() {
        val handler = Handler(Looper.getMainLooper())
        val runnable = Runnable {
            val intentArray = arrayOfNulls<Intent>(2)
            intentArray[0] = Intent(context, MainActivity::class.java)
            intentArray[0]?.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
            intentArray[1] = Intent(context, LaunchBAPMActivity::class.java)
            context.startActivities(intentArray)
        }
        handler.postDelayed(runnable, 750)
    }

    fun launchDisconnectActivity() {
        val launchIntent = Intent(context, DisconnectActivity::class.java).also {
            it.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        context.startActivity(launchIntent)
    }

    fun canMapsLaunchNow(): Boolean {
        setDirectionLocation()
        val isLaunchingWithDirections = preferencesHelper.isLaunchingWithDirections
        val isUsingTimesToLaunch = preferencesHelper.isUsingTimesToLaunch
        return isLaunchingWithDirections || isUsingTimesToLaunch
    }

    private fun setDirectionLocation() {
        if (canLaunchDuringThisTime()) {
            directionLocation = canLaunchThisTimeLocations.find { directionLocation ->
                canLaunchOnThisDay(directionLocation)
            } ?: NONE
        }
        canLaunchThisTimeLocations.clear()
    }

    fun canLaunchOnThisDay(directionLocation: DirectionLocation): Boolean {
        val calendar = Calendar.getInstance()
        val today = calendar.get(Calendar.DAY_OF_WEEK).toString()

        val daysToLaunchHome = preferencesHelper.daysToLaunchHome
        val daysToLaunchWork = preferencesHelper.daysToLaunchWork
        val daysToLaunchCustom = preferencesHelper.daysToLaunchCustom

        val canLaunch = when (directionLocation) {
            HOME -> daysToLaunchHome.contains(today)
            WORK -> daysToLaunchWork.contains(today)
            CUSTOM -> daysToLaunchCustom.contains(today)
            NONE -> false
        }

        Log.d(TAG, "Day of the week: $today")
        Log.d(TAG, "Can Launch: $canLaunch")
        Log.d(TAG, "Direction Location: $directionLocation")

        return canLaunch
    }

    fun canLaunchDuringThisTime(): Boolean {
        val isUseLaunchTimeEnabled = preferencesHelper.isUseLaunchTimeEnabled
        if (isUseLaunchTimeEnabled) {
            val current24hrTime = TimeHelper.current24hrTime
            val morningStartTime = preferencesHelper.morningStartTime
            val morningEndTime = preferencesHelper.morningEndTime
            val eveningStartTime = preferencesHelper.eveningStartTime
            val eveningEndTime = preferencesHelper.eveningEndTime
            val customStartTime = preferencesHelper.customStartTime
            val customEndTime = preferencesHelper.customEndTime

            val timeHelperMorning = TimeHelper(morningStartTime, morningEndTime, current24hrTime)
            val canLaunchWork = timeHelperMorning.isWithinTimeSpan
            if (canLaunchWork) {
                canLaunchThisTimeLocations.add(WORK)
            }

            val timeHelperEvening = TimeHelper(eveningStartTime, eveningEndTime, current24hrTime)
            val canLaunchHome = timeHelperEvening.isWithinTimeSpan
            if (canLaunchHome) {
                canLaunchThisTimeLocations.add(HOME)
            }

            val timeHelperCustom = TimeHelper(customStartTime, customEndTime, current24hrTime)
            val canLaunchCustom = timeHelperCustom.isWithinTimeSpan
            if (canLaunchCustom) {
                canLaunchThisTimeLocations.add(CUSTOM)
            }

            return canLaunchWork || canLaunchHome || canLaunchCustom
        } else {
            return true
        }
    }

    fun launchWazeDirections(location: String) {
        val handler = Handler()
        val runnable = Runnable {
            val uriString = "waze://?favorite=$location&navigate=yes"
            Log.d(TAG, "DIRECTIONS LOCATION: $uriString")
            val uri = Uri.parse(uriString)
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = uri
            packageHelper.sendBroadcast(intent)
        }

        handler.postDelayed(runnable, 4000)
    }

    fun closeWazeOnDisconnect() {
        val handler = Handler()
        val runnable = Runnable {
            val intent = Intent()
            intent.action = "Eliran_Close_Intent"
            packageHelper.sendBroadcast(intent)
        }

        handler.postDelayed(runnable, 2000)
    }

    fun isAbleToLaunch(packageName: String) = packageHelper.isPackageOnPhone(packageName)

    companion object {
        private const val TAG = "LaunchAppHelper"
    }
}

